specs/language/lex.tex

```diff
@@ -0,0 +1,48 @@
+\Ch{Lexical Conventions}{Lex}
+
+\Sec{Unit of Translation}{Lex.Translation}
+
+\p The text of \acrshort{hlsl} programs is collected in \textit{source} and
+\textit{header} files. The distinction between source and header files is social
+and not technical. An implementation will construct a \textit{translation unit}
+from a single source file and any included source or header files referenced via
+the \texttt{\#include} preprocessing directive conforming to the \gls{isoC}
+preprocessor specification.
+
+\p An implementation may implicitly include additional sources as required to
+expose the \acrshort{hlsl} library functionality as defined in (FIXME: Add
+reference to library chapter).
+
+\Sec{Phases of Translation}{Lex.Phases}
+
+\p \acrshort{hlsl} inherits the phases of translation from \gls{isoCPP}, with
+minor alterations, specifically the removal of support for trigraph and digraph
+sequences. Below is a description of the phases.
+
+\begin{enumerate}
+  \item Source files are characters are mapped to the basic source character set
+  in an implementation-defined manner.
+  \item Any sequence of backslash (\texttt{\textbackslash}) immediately followed
+  by a new line is deleted, resulting in splicing lines together.
+  \item Tokenization occurs and comments are isolated. If a source file ends in
+  a partial comment or preprocessor token the program is ill-formed and a
+  diagnostic shall be issued. Each comment block shall be treated as a single
+  white-space character.
+  \item Preprocessing directives are executed, macros are expanded,
+  \texttt{pragma} and other unary operator expressions are executed. Processing
+  of \texttt{\#include} directives results in all preceding steps being executed
+  on the resolved file. Finally all preprocessing directives are removed from
```

Comment:
Is it worth mentioning that the process of processing include directives is recursive? I should also note that there isn't one final step to remove all preprocessing directives from the source, rather, (at least for includes), the included source gets preprocessed and injected into the parent source file, and the include directive gets removed in a step by step manner. Is this worth mentioning?

Reply:
The compiler doesn't really inject the preprocessed source into the parent source file during normal compilation. Compilers generally support that for emitting preprocessed source, but during actual compilation the expansion occurs usually on the token iterators not by merging memory buffers to create a preprocessed file in memory.

The specification here is intended to be vague on this point so that the implementation is free to optimize as it sees fit.